File: app/src/main/java/hanas/aptacy/moodlamp/MainActivityPresenter.java
package hanas.aptacy.moodlamp;

import android.app.WallpaperManager;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.util.Log;

import com.spotify.android.appremote.api.ConnectionParams;
import com.spotify.android.appremote.api.Connector;
import com.spotify.android.appremote.api.SpotifyAppRemote;
import com.spotify.protocol.types.PlayerState;
import com.spotify.protocol.types.Track;

import java.io.InputStream;
import java.util.List;
import java.util.concurrent.TimeUnit;

import hanas.aptacy.moodlamp.services.LeadingColorsService;
import hanas.aptacy.moodlamp.services.pojo.LeadingColorBody;
import okhttp3.OkHttpClient;
import okhttp3.ResponseBody;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class MainActivityPresenter {

    private static final String LEADING_COLORS_URL = "https://moodlamp-leading-color.herokuapp.com";
    private static final String CLIENT_ID = "6d7d697321a74a419a8ede8b74c6a7c8";
    private static final String REDIRECT_URI = "https://trello.com/b/zYzvg43L/moodlamp";

    private SpotifyAppRemote spotifyAppRemote;
    private LeadingColorsService leadingColorsService;
    private Track previousTrack;

    private MainActivityView view;

    public MainActivityPresenter(MainActivityView view) {
        this.view = view;
    }

    public void connectWithLeadingColorsService() {
        HttpLoggingInterceptor logging = new HttpLoggingInterceptor();
        logging.setLevel(HttpLoggingInterceptor.Level.BODY);

        OkHttpClient.Builder httpClient = new OkHttpClient.Builder()
                .readTimeout(30, TimeUnit.SECONDS)
                .writeTimeout(30, TimeUnit.SECONDS);

        httpClient.addInterceptor(logging);  // <-- this is the important line!

        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl(LEADING_COLORS_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .client(httpClient.build())
                .build();

        leadingColorsService = retrofit.create(LeadingColorsService.class);
    }

    private class SpotifyConnectionListener implements Connector.ConnectionListener {
        @Override
        public void onConnected(SpotifyAppRemote mSpotifyAppRemote) {
            spotifyAppRemote = mSpotifyAppRemote;
            Log.d("MainActivity", "Connected! Yay!");

            // Now you can start interacting with App Remote
            spotifyAppRemote.getPlayerApi()
                    .subscribeToPlayerState()
                    .setEventCallback(MainActivityPresenter.this::newSpotifySong);
        }

        @Override
        public void onFailure(Throwable throwable) {
            view.showSpotifyConnectionError(throwable);
            Log.e("MainActivity", throwable.getMessage(), throwable);

            // Something went wrong when attempting to connect! Handle errors here
        }
    }

    public void connectWithSpotifyAppRemote() {
        // Set the connection parameters
        ConnectionParams connectionParams =
                new ConnectionParams.Builder(CLIENT_ID)
                        .setRedirectUri(REDIRECT_URI)
                        .showAuthView(true)
                        .build();

        view.connectWithSpotify(connectionParams, new SpotifyConnectionListener());
    }


    private void newSpotifySong(PlayerState playerState) {
        final Track track = playerState.track;
        if (track != null && (previousTrack == null || !track.imageUri.equals(previousTrack.imageUri))) {
            String imageUri = track.imageUri.raw;
            String imageId = imageUri.substring(imageUri.lastIndexOf(":") + 1);
            String imageURLString = "https://i.scdn.co/image/" + imageId;
            view.setSongInfo(track.name, track.artist.name, imageURLString);
            Call<List<LeadingColorBody>> call = leadingColorsService.getColors(imageURLString);
            call.enqueue(new Callback<List<LeadingColorBody>>() {
                @Override
                public void onResponse(Call<List<LeadingColorBody>> call, Response<List<LeadingColorBody>> response) {
                    List<LeadingColorBody> leadingColors = response.body();
                    view.setColorsOfNewAlbumCover(leadingColors);
                    int[] screenSize = view.getScreenSize();
                    Call<ResponseBody> callImage = leadingColorsService.getWallpaper(10000, screenSize[1], screenSize[0], leadingColors);
                    callImage.enqueue(new Callback<ResponseBody>() {
                        @Override
                        public void onResponse(Call<ResponseBody> call, Response<ResponseBody> response) {
                            InputStream is = response.body().byteStream();
                            Bitmap bitmap = BitmapFactory.decodeStream(is);
                            view.setWallpaperFromBitmap(bitmap);
                        }

                        @Override
                        public void onFailure(Call<ResponseBody> call, Throwable t) {
                            view.showLeadingColorServiceResponseError(t);
                            Log.e("API failure", t.getMessage(), t);
                        }
                    });
                    Log.d("API response", response.toString());
                }

                @Override
                public void onFailure(Call<List<LeadingColorBody>> call, Throwable t) {
                    view.showLeadingColorServiceResponseError(t);
                    Log.e("API failure", t.getMessage(), t);
                }
            });
            previousTrack = track;
        }
    }
}

File: app/src/main/java/hanas/aptacy/moodlamp/MainActivityPresenter.kt
package hanas.aptacy.moodlamp

import android.graphics.BitmapFactory
import hanas.aptacy.moodlamp.services.leadingcolors.LeadingColorsService
import hanas.aptacy.moodlamp.services.leadingcolors.pojo.LeadingColorBody
import hanas.aptacy.moodlamp.services.spotify.SpotifyService
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import okhttp3.Dispatcher
import okhttp3.ResponseBody
import javax.inject.Inject

class MainActivityPresenter @Inject constructor(
        private val view: MainActivityView,
        private val dispatcher: Dispatcher,
        private val leadingColorsService: LeadingColorsService,
        private val scope: CoroutineScope
) : SpotifyService.Callback {

    private var previousTrackImageUri: String? = null

    @Inject
    fun runSpotifyService(spotifyService: SpotifyService) {
        spotifyService.connect(this)
    }

    private suspend fun getColorsSuccessful(response: List<LeadingColorBody?>?) {
        val responseBody = response ?: return
        val leadingColors: List<LeadingColorBody> = responseBody.filterNotNull()
        val prominentColor: LeadingColorBody = leadingColors
                .filter { leadingColorBody: LeadingColorBody -> leadingColorBody.hsv[2] > 0.2 && leadingColorBody.hsv[2] < 0.8 }
                .maxByOrNull { leadingColorBody: LeadingColorBody -> leadingColorBody.hsv[1] }
                ?: leadingColors[0]

        withContext(Dispatchers.Main) {
            view.setFabColor(prominentColor.getColorInt())
            view.setColorsOfNewAlbumCover(leadingColors)
            val screenSize: IntArray = view.screenSize ?: intArrayOf(2000, 1000)
            withContext(Dispatchers.IO) {
                val response = leadingColorsService.getWallpaper(2000, screenSize[1], screenSize[0], leadingColors)
                getWallpaperSuccessful(response)
            }
        }
    }

    private suspend fun getWallpaperSuccessful(response: ResponseBody?) {
        val responseBody = response ?: return
        val bitmap = BitmapFactory.decodeStream(responseBody.byteStream())
        withContext(Dispatchers.Main) {
            view.setWallpaperToActivityBackground(bitmap)
            view.setWallpaperFromBitmap(bitmap)
        }
    }

    private fun serviceFailure(throwable: Throwable) {
        view.showLeadingColorServiceResponseError(throwable)
    }

    override fun onSpotifyConnected() = Unit

    override fun onSpotifyConnectionFailure(throwable: Throwable?) = view.showSpotifyConnectionError(throwable)

    private fun isNewWallpaperNeeded(trackImageUri: String?) = trackImageUri != previousTrackImageUri

    override fun onNewSpotifyTrack(trackName: String?, artist: String?, coverUrl: String?) {
//        if (!this::dispatcher.isInitialized || !this::leadingColorsService.isInitialized) return

        if (isNewWallpaperNeeded(coverUrl)) {
            dispatcher.cancelAll()
            view.setSongInfo(trackName, artist, coverUrl)
            previousTrackImageUri = coverUrl
            scope.launch {
                try {
                    val response = leadingColorsService.getColors(coverUrl)
                    getColorsSuccessful(response)
                } catch (t: Throwable) {
                    withContext(Dispatchers.Main) {
                        serviceFailure(t)
                    }
                }
            }
        }
    }
}